### ISSUE TITLE
F841 false positive in def-for-if-else

### ISSUE THREAD

**Issue by gamer191** [astral-sh/ruff#28324]
### Summary

https://play.ruff.rs/b5126a27-8289-4c5e-8bb5-ed6817f87d86

```
def c():
    for a in [1,2]:
        if a == 2:
            print(b)
        else:
            b=3
```
incorrectly reports b as "Local variable `b` is assigned to but never used (F841) "

### Version

Ruff Playground

**Comment by ntBre** [comment]
Thanks for the report! However, I don't think I'd really consider this a false positive from Ruff. In this very specific case, it's "obvious" that `b` is initialized and later used because the list is a literal, and the condition in the `if` also checks a literal, but this quickly gets complicated if the iterator comes from somewhere else or the condition calls a function from another module, etc.

Even the Rust compiler doesn't allow this pattern:

```rust
fn main() {
    let mut b;
    for a in [1, 2] {
        if a == 2 {
            println!("{}", b);
        } else {
            b = 3;
        }
    }
}
```

```
error[E0381]: used binding `b` is possibly-uninitialized
 --> src/main.rs:5:24
  |
2 | let mut b;
  |     ----- binding declared here but left uninitialized
3 | for a in [1, 2] {
  |          ------ if the `for` loop runs 0 times, `b` is not initialized
4 |     if a == 2 {
  |        ------ if this condition is `true`, `b` is not initialized
5 |         println!("{}", b);
  |                        ^ `b` used here but it is possibly-uninitialized
  |
  = note: when checking initialization, the compiler describes possible control-flow paths without evaluating whether branch conditions can actually have the values shown
```

That's the compiler's variant of F821 rather than F841, but the two rules are aspects of the same underlying issue, which is tracked in
https://github.com/astral-sh/ruff/issues/12451. Some aspects of this might be improved with better https://github.com/astral-sh/ruff/labels/type-inference.
